app/controllers/two_factor_authentication/otp_expired_controller.rb

```diff
@@ -0,0 +1,26 @@
+module TwoFactorAuthentication
+  class OtpExpiredController < ApplicationController
+    before_action :confirm_two_factor_authenticated, unless: :user_signed_in?
+    before_action :validate_otp_expiration
+
+    def show
+      @otp_delivery_preference = otp_delivery_preference
+    end
+
+    private
+
+    def otp_delivery_preference
+      current_user.otp_delivery_preference
+    end
+
+    def validate_otp_expiration
+      redirect_to after_sign_in_path_for(current_user) if user_signed_in? && !otp_expired?
+    end
+
+    def otp_expired?
+      return if current_user.direct_otp_sent_at.blank?
+      (current_user.direct_otp_sent_at +
+        TwoFactorAuthenticatable::DIRECT_OTP_VALID_FOR_SECONDS).past?
+    end
```

Comment:
When testing this, I ran into an issue where when I was redirected, I never saw the expired screen, it looked as though the "Enter your one-time code" page just refreshed. I suspect there could be a race condition where the user is redirected at almost the exact time that the OTP becomes expired, so this enforcement of `otp_expired?` may not always result in `true` as expected.

I wonder if we should just drop this validation? The bigger issue discussed in https://github.com/18F/identity-idp/pull/7577/files#r1082910213 was mostly about avoiding a 500 error because the code references a `current_user` which may not necessarily exist. I think as long as a user exists, I don't think it would be a problem if the user navigated to the page directly. What do you think?

```suggestion
```

Alternatively, we could change it to just check that an OTP has been sent, but not concern too much about whether it's expired or not.

```suggestion
    def validate_otp_expiration
      redirect_to after_sign_in_path_for(current_user) if user_signed_in? && current_user.direct_otp_sent_at.blank?
    end
```

Reply:
Dropping the validation makes the most sense to me. This way, if someone hit "Send code" on that screen when accessed directly, it will be as if you request a code for the first time.